### PR TITLE
Add jackline and (updated) ocaml-dependencies

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Addition of `jackline`, a terminal-based XMPP client purely in OCaml.

For `jackline` added missing dependencies:

* `ptime`
* `tls`
* `notty`
* `otr`
* `erm_xmpp_0_3`

and updated the following packages:

* `x509`
* `cstruct`
* `nocrypto`
* `uutf`
* `uuseg`

This PR also incudes minor fixes for:

* ~~~`zarith`~~~ (@vbgl has fixed it on master)
* `ppx_conv_sexp`

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] OS X
   - [ ] Linux
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Builds:

* [x] `ocamlPackages`
* [x] `ocamlPackages_4_02`
* [ ] `ocamlPackages_4_03`

**Comment by vbgl** [comment]
You may want to look at the derivation for `topkg`, and make your `buildPhase` and `installPhase` similar to its: use `-I …` instead of a symlink; move (or link) the result directory to where it is expected by the nixpkgs version of `findlib`.

**Comment by sternenseemann** [comment]
@vbgl Thank you, I will have a look at it.

**Comment by sternenseemann** [comment]
@vbgl I updated it. That is clearly better.

**Comment by vbgl** [comment]
The `installPhase` of `ptime` is not correct yet: you need to move or symlink the `$out/lib/ptime` directory to `$out/lib/ocaml/${ocaml.version}/site-lib`; otherwise, you won’t be able to use that package (at least, `ocamlfind` won’t find it).

**Comment by sternenseemann** [comment]
@vbgl Thank you, I fixed it. Is this only an issue if using `opam-installer`?

**Comment by sternenseemann** [comment]
Note: This PR depends on #20086, otherwise `opam` does not build.

File: pkgs/development/ocaml-modules/x509/default.nix [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
   doCheck = true;
   checkTarget = "test";
   createFindlibDestdir = true;
 
   meta = {
     homepage = https://github.com/mirleft/ocaml-x509;
     description = "X509 (RFC5280) handling in OCaml";
-    platforms = ocaml.meta.platforms or [];
-    license = stdenv.lib.licenses.bsd2;
+    license = stdenv.licenses.bsd2;
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/x509/default.nix]: stdenv.licenses → stdenv.lib.licenses

**Review by Profpatsch — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/jackline/default.nix [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      --pinned true
+  '';
+
+  installPhase = ''
+    opam-installer --prefix=$out --script | sh
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/hannesm/jackline;
+    description = "terminal-based XMPP client purely in OCaml";
```
Comment (Profpatsch) [inline pkgs/applications/networking/instant-messengers/jackline/default.nix]: All descriptions start with a capital letter and end with a `.`.

File: pkgs/development/ocaml-modules/astring/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  '';
+
+  installPhase = ''
+    opam-installer --script --prefix=$out astring.install | sh
+    ln -s $out/lib/astring $out/lib/ocaml/${ocaml.version}/site-lib/
+  '';
+
+  meta = {
+    homepage = http://erratique.ch/software/ptime;
+    description = "Alternative String module for OCaml";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/astring/default.nix]: `.`

File: pkgs/development/ocaml-modules/cstruct/default.nix [hunk 4]
```diff
@@ -1,20 +1,29 @@
-{stdenv, writeText, fetchurl, ocaml, ocplib-endian, sexplib_p4, findlib,
- async_p4 ? null, lwt ? null, camlp4}:
+{stdenv, buildOcaml, fetchFromGitHub, writeText, ocaml, ocplib-endian, sexplib_p4, findlib,
+ async_p4 ? null, lwt ? null, camlp4, ppx_tools ? null}:
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: It’ a lot nicer to explicitly create switches for those, then sane defaults can be enforced directly in the package function:

```nix
{ asyncSupport ? false, async_p4 ? null
, … }:

with stdenv.lib;
assert asyncSupport -> async_p4 != null;
…
  configureFlags = optional asyncSupport async_p4 ++ …
```

File: pkgs/development/ocaml-modules/cstruct/default.nix [hunk 5]
```diff
@@ -1,20 +1,29 @@
...
+                   (if lwt != null then ["--enable-lwt"] else []) ++
+                   (if async_p4 != null then ["--enable-async"] else []) ++
+                   (if ppx_tools != null then ["--enable-ppx"] else []);
+
   buildInputs = [ocaml findlib camlp4];
-  propagatedBuildInputs = [ocplib-endian sexplib_p4 lwt async_p4];
+  propagatedBuildInputs = [ocplib-endian sexplib_p4 lwt async_p4 ppx_tools];
+
+  doCheck = true;
+  checkTarget = "test";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: needed?

File: pkgs/development/ocaml-modules/erm_xmpp/0.3.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  configurePhase = "ocaml setup.ml -configure --prefix $out";
+  buildPhase = "ocaml setup.ml -build";
+  installPhase = "ocaml setup.ml -install";
+
+  createFindlibDestdir = true;
+
+  meta = {
+    homepage = https://github.com/hannesm/xmpp;
+    description = "OCaml based XMPP implementation (fork)";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/erm_xmpp/0.3.nix]: .

File: pkgs/development/ocaml-modules/nocrypto/default.nix [hunk 7]
```diff
@@ -1,28 +1,33 @@
-{ stdenv, fetchzip, ocaml, findlib, cstruct, type_conv, zarith, ounit }:
+{ stdenv, buildOcaml, fetchFromGitHub, ocaml, findlib, cstruct, type_conv, zarith, ounit, ocaml_oasis, ppx_sexp_conv, lwt ? null}:
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/nocrypto/default.nix]: see above cf. `null`

File: pkgs/development/ocaml-modules/notty/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
+{stdenv, buildOcaml, fetchFromGitHub, findlib, result, uucp, uuseg, uutf, lwt ? null}:
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/notty/default.nix]: `null`

File: pkgs/development/ocaml-modules/x509/default.nix (hunk 1 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/x509/default.nix]: Fixed, thank you!

File: pkgs/development/ocaml-modules/ptime/default.nix [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  buildInputs = [ ocaml findlib ocamlbuild topkg opam ];
+
+  propagatedBuildInputs = [ result ];
+
+  buildPhase = ''
+    # this is a tiny hack to make #use "topfind" work.
+    # the hack used in ocamlPackage.containers didn't work here…
+    ln -s ${mkpath findlib}/topfind topfind
+    ocaml pkg/pkg.ml build --with-js_of_ocaml false
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/ptime/default.nix]: Fixed. It is possible with `-I` as @vbgl pointed out earlier.

File: pkgs/development/ocaml-modules/uutf/default.nix [hunk 10]
```diff
@@ -1,32 +1,38 @@
-{ stdenv, fetchurl, ocaml, findlib, ocamlbuild, opam }:
+{ stdenv, buildOcaml, fetchurl, ocaml, findlib, ocamlbuild, opam, cmdliner}:
 let
   pname = "uutf";
   webpage = "http://erratique.ch/software/${pname}";
 in
 
-assert stdenv.lib.versionAtLeast ocaml.version "3.12";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/uutf/default.nix]: `minimumSupportedOcamlVersion`

File: pkgs/applications/networking/instant-messengers/jackline/default.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
+{stdenv, fetchFromGitHub, ocamlPackages}:
+
+assert stdenv.lib.versionAtLeast ocamlPackages.ocaml.version "4.02.2";
```
Comment (Profpatsch) [inline pkgs/applications/networking/instant-messengers/jackline/default.nix]: `minimumSupportedOcamlVersion`

File: pkgs/development/ocaml-modules/cstruct/default.nix [hunk 12]
```diff
@@ -1,20 +1,29 @@
-{stdenv, writeText, fetchurl, ocaml, ocplib-endian, sexplib_p4, findlib,
- async_p4 ? null, lwt ? null, camlp4}:
+{stdenv, buildOcaml, fetchFromGitHub, writeText, ocaml, ocplib-endian, sexplib_p4, findlib,
+ async_p4 ? null, lwt ? null, camlp4, ppx_tools ? null}:
 
 assert stdenv.lib.versionAtLeast (stdenv.lib.getVersion ocaml) "4.01";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: `minimumSupportedOcamlVersion`

File: pkgs/development/ocaml-modules/nocrypto/default.nix [hunk 13]
```diff
@@ -1,28 +1,33 @@
...
 
 assert stdenv.lib.versionAtLeast ocaml.version "4.01";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/nocrypto/default.nix]: `minimumSupportedOcamlVersion`

File: pkgs/development/ocaml-modules/notty/default.nix [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+  src = fetchFromGitHub {
+    owner  = "pqwy";
+    repo   = "notty";
+    rev    = "v${version}";
+    sha256 = "0bw3bq8z2y1rhc20zn13s78sazywyzpg8nmyjch33p7ypxfglf01";
+  };
+
+  buildInputs = [ findlib ];
+  propagatedBuildInputs = [ lwt result uucp uuseg uutf ];
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/notty/default.nix]: Aren’t those normal ocamlPackages?

File: pkgs/development/ocaml-modules/notty/default.nix [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+  configureFlags = [ "--enable-unix" ] ++
+                   (if lwt != null then [ "--enable-lwt" ] else []);
+  configurePhase = "./configure --prefix $out $configureFlags";
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/pqwy/notty/tree/master;
+    description = "Declarative terminal graphics for OCaml";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/notty/default.nix]: .

File: pkgs/development/ocaml-modules/otr/default.nix [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    ocaml ${ocamlFlags}  pkg/pkg.ml build \
+      --tests true
+  '';
+
+  installPhase = ''
+    opam-installer --prefix=$out --script | sh
+    ln -s $out/lib/otr $out/lib/ocaml/${ocaml.version}/site-lib
+  '';
+
+  doCheck = true;
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/otr/default.nix]: needed?

File: pkgs/development/ocaml-modules/otr/default.nix [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  checkPhase = "ocaml ${ocamlFlags} pkg/pkg.ml test";
+
+  createFindlibDestdir = true;
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/hannesm/ocaml-otr;
+    description = "Off-the-record (OTR) messaging protocol, purely in OCaml";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/otr/default.nix]: .

File: pkgs/development/ocaml-modules/ptime/default.nix [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  src = fetchurl {
+    url = "http://erratique.ch/software/ptime/releases/ptime-${version}.tbz";
+    sha256 = "1lihkhzskzwxskiarh4mvf7gbz5nfv25vmazbfz81m344i32a5pj";
+  };
+
+  unpackCmd = "tar -xf $curSrc";
+
+  buildInputs = [ ocaml findlib ocamlbuild topkg opam ];
+
+  propagatedBuildInputs = [ result ];
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/ptime/default.nix]: buildInputs?

File: pkgs/development/ocaml-modules/ptime/default.nix [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+  '';
+
+  installPhase = ''
+    opam-installer --script --prefix=$out ptime.install | sh
+    ln -s $out/lib/ptime $out/lib/ocaml/${ocaml.version}/site-lib
+  '';
+
+  meta = {
+    homepage = http://erratique.ch/software/ptime;
+    description = "POSIX time for OCaml";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/ptime/default.nix]: .

File: pkgs/development/ocaml-modules/tls/default.nix [hunk 20]
```diff
@@ -0,0 +1,32 @@
+{stdenv, buildOcaml, fetchFromGitHub, findlib, ocamlbuild, ocaml_oasis, ppx_tools, ppx_sexp_conv, result, x509, nocrypto, cstruct, lwt ? null}:
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/tls/default.nix]: `null`

File: pkgs/development/ocaml-modules/tls/default.nix [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+  buildInputs = [ ocamlbuild findlib ocaml_oasis ppx_sexp_conv ];
+  propagatedBuildInputs = [ cstruct lwt nocrypto result x509 ];
+
+  configureFlags = [ "--disable-mirage" "--enable-tests" ] ++
+                   (if lwt != null then [ "--enable-lwt" ] else [ "--disable-lwt" ]);
+
+  configurePhase = "./configure --prefix $out $configureFlags";
+
+  doCheck = true;
+  checkTarget = "test";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/tls/default.nix]: needed?

File: pkgs/development/ocaml-modules/tls/default.nix [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+  createFindlibDestdir = true;
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/mirleft/ocaml-tls;
+    description = "TLS in pure OCaml";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/tls/default.nix]: .

File: pkgs/development/ocaml-modules/uuseg/default.nix [hunk 23]
```diff
@@ -1,40 +1,39 @@
...
 in
 
-assert versionAtLeast (getVersion ocaml) "4.01";
+buildOcaml rec {
 
-stdenv.mkDerivation {
+  minimumSupportedOcamlVersion = "4.01";
 
-  name = "ocaml-${pname}-${version}";
+  name = "${pname}";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/uuseg/default.nix]: name = pname;

File: pkgs/development/ocaml-modules/uutf/default.nix [hunk 24]
```diff
@@ -1,32 +1,38 @@
...
+buildOcaml rec {
+  name = "${pname}";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/uutf/default.nix]: `name = pname;`

File: pkgs/development/ocaml-modules/x509/default.nix [hunk 25]
```diff
@@ -1,28 +1,30 @@
...
+    rev    = "${version}";
+    sha256 = "07cc3z6h87460z3f4vz8nlczw5jkc4vjhix413z9x6nral876rn7";
   };
 
-  buildInputs = [ ocaml findlib ounit ];
+  buildInputs = [ ocaml ocaml_oasis findlib ounit ppx_sexp_conv ];
   propagatedBuildInputs = [ asn1-combinators nocrypto ];
 
   configureFlags = "--enable-tests";
+  configurePhase = "./configure --prefix $out $configureFlags";
```
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/x509/default.nix]: Isn’t the prefix set automatically?

**Review by Profpatsch — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/cstruct/default.nix (hunk 5 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: Yes.

File: pkgs/development/ocaml-modules/notty/default.nix (hunk 14 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/notty/default.nix]: Could you elaborate a bit?

File: pkgs/development/ocaml-modules/uutf/default.nix (hunk 10 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/uutf/default.nix]: Is already used here.

File: pkgs/development/ocaml-modules/ptime/default.nix (hunk 18 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/ptime/default.nix]: No, is correct like this.

File: pkgs/development/ocaml-modules/tls/default.nix (hunk 21 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/tls/default.nix]: Yes.

File: pkgs/development/ocaml-modules/x509/default.nix (hunk 25 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/x509/default.nix]: Yes, but as `--prefix=$out` which `./configure` of this package can not parse.

File: pkgs/development/ocaml-modules/cstruct/default.nix (hunk 4 above)
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: Good point

File: pkgs/development/ocaml-modules/cstruct/default.nix (hunk 5 above)
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/cstruct/default.nix]: doCheck as well?

File: pkgs/development/ocaml-modules/notty/default.nix (hunk 14 above)
Comment (Profpatsch) [inline pkgs/development/ocaml-modules/notty/default.nix]: Was about the propagated vs normal build inputs, apparently ocaml needs some packages propagated.

Comment (sternenseemann) [inline pkgs/development/ocaml-modules/notty/default.nix]: Yeah, indeed. If a package's interface is used in the module, it needs the package propagated, I guess.

**Comment by sternenseemann** [comment]
@vbgl What do you think, do we still have to keep the checks for `lwt`? It seems to build fine with all ocaml versions.

**Review by Profpatsch — APPROVED** [review]

**Comment by vbgl** [comment]
Reverting commit  32e86a3 fixes the `core_kernel` build. However, this breaks `nocrypto-0.5.3`. Should we revert this whole PR, which apparently deserves more testing?

**Comment by Profpatsch** [comment]
@vbgl yes, let’s do that. OcamlPackages is still in an extremely unstable state.

File: pkgs/applications/networking/instant-messengers/jackline/default.nix (hunk 2 above)
Comment (vbgl) [inline pkgs/applications/networking/instant-messengers/jackline/default.nix]: [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) reads: “`meta.description` should […] Not have a dot at the end”

**Comment by vbgl** [comment]
@sternenseemann would you like to open a new PR, or preferably (IMHA) several PRs, so that we can review and test independent changes independently.

Concerning `ptime` two nitpicking remarks: you don’t need to define `ocaml-version` as you can directly refer to `ocaml.version`; the `meta.description` should not end with a dot.

**Comment by sternenseemann** [comment]
@vbgl sounds like a plan!

**Pull request merged**